private/pkg/manifest/module_test.go

```diff
@@ -168,6 +171,208 @@ func TestBlobFromReader(t *testing.T) {
 	)
 }
 
+type mockBlob struct {
+	digest  *manifest.Digest
+	content io.Reader
+	openErr bool
+}
+
+func (mb *mockBlob) Digest() *manifest.Digest { return mb.digest }
+func (mb *mockBlob) Open(_ context.Context) (io.ReadCloser, error) {
+	if mb.openErr {
+		return nil, errors.New("open error")
+	}
+	return io.NopCloser(mb.content), nil
+}
+
+type errAtEndReader struct{ reader io.Reader }
+
+func (e *errAtEndReader) Read(p []byte) (int, error) {
+	n, err := e.reader.Read(p)
+	if err == io.EOF {
+		return n, errors.New("test erroring at EOF")
+	}
+	return n, err
+}
+
+func TestBlobEqual(t *testing.T) {
+	testBlobEqual(
+		t,
+		"does equal",
+		strings.NewReader("foo"),
+		strings.NewReader("foo"),
+		true,
+		false,
+	)
+	testBlobEqual(
+		t,
+		"equal with differing length reads",
+		strings.NewReader("foo"),
+		iotest.OneByteReader(strings.NewReader("foo")),
+		true,
+		false,
+	)
+	testBlobEqual(
+		t,
+		"mismatched equal-length content",
+		strings.NewReader("foo"),
+		strings.NewReader("bar"),
+		false,
+		false,
+	)
+	testBlobEqual(
+		t,
+		"mismatched equal-length content data with error",
+		iotest.DataErrReader(strings.NewReader("foo")),
+		iotest.DataErrReader(strings.NewReader("bar")),
+		false,
+		false,
```

Comment:
Don't fully understand why this one doesn't return error.

Reply:
DataErrReader forces strings.Reader to return n > 0 with error (io.EOF). io.EOF isn't an real error in the sense that it's a stream successfully terminated signal. That's why the data is different (false / does-not-equal) but the io.EOF from the paired read is absorbed (successfully reached end of both streams).

To help make sense, this test activates the bytes.Equal path (they don't) yet return no error because of nilEOF filtering.

```go
if !bytes.Equal(aBlock[:aN], bBlock[:bN]) {
	// Read content doesn't match. Don't forget to forward any errors
	// returned.
	return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
}
```